CLDR-16972 CLDRModify -fQ is messing up some annotations

btangmu opened the pull request:
-Call getDistinguishingXPath in CLDRModify to avoid getting bogus code-fallback

-Enhance DisplayAndInputProcessor.filterCoveredKeywords so PANDA BEAR matches panda and bear

-Add a unit test TestFilterCoveredKeywords

-TODO comment, still need retroactive fix for japanse poskantoor | Japanse poskantoor

CLDR-16972

- [ ] This PR completes the ticket.

<!--
Thank you for your pull request.
Please see https://cldr.unicode.org/index/process for general
information on contributing to CLDR.

1. Make sure the ticket is filed at
https://unicode-org.atlassian.net/projects/CLDR/
2. Update the PR title and first line of this
message to include the ticket ID (CLDR-_____)
3. You will be automatically asked to sign the contributors’
license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/cldr
- license: https://www.unicode.org/copyright.html#License
-->

ALLOW_MANY_COMMITS=true

btangmu commented on tools/cldr-code/src/main/java/org/unicode/cldr/test/DisplayAndInputProcessor.java:
```diff
@@ -18,14 +18,7 @@
-import java.util.Map;
-import java.util.Set;
-import java.util.TreeSet;
+import java.util.*;
```
I ran this and it doesn't cause any changes:
`mvn --file=tools/pom.xml spotless:apply`

Is there an additional formatter I should be running?

srl295 replied:
indeed there are 32 files with `import org.unicode.cldr.util.*;` in the source

btangmu replied:
https://stackoverflow.com/questions/3348816/intellij-never-use-wildcard-imports#:~:text=The%20solution%20is%20to%20go%20to%20Preferences%20%28%E2%8C%98,Any%20value%20over%2099%20seems%20to%20work%20fine.

btangmu commented on tools/cldr-code/src/main/java/org/unicode/cldr/test/DisplayAndInputProcessor.java:
```diff
@@ -686,21 +679,34 @@ private static String annotationsForDisplay(String value) {
             if (sorted.containsAll(list)) {
                 toRemove.add(item);
+            } else {
+                List<String> listLower = new ArrayList<>();
```
I enlarged the unit test as follows:
```
    public void TestFilterCoveredKeywords() {
        TreeSet<String> set = new TreeSet<>();
        set.add("bear");
        set.add("panda");
        set.add("panda bear");
        DisplayAndInputProcessor.filterCoveredKeywords(set);
        if (set.contains("panda bear")) {
            errln("panda bear should be filtered out");
        } else {
            set.add("PANDA BEAR");
            DisplayAndInputProcessor.filterCoveredKeywords(set);
            if (set.contains("PANDA BEAR")) {
                errln("PANDA BEAR should be filtered out");
            }
        }
        set.clear();
        set.add("bEAR");
        set.add("Panda");
        set.add("panda bear");
        DisplayAndInputProcessor.filterCoveredKeywords(set);
        if (set.contains("panda bear")) {
            errln("panda bear should be filtered out");
        } else {
            set.add("PandA beAR");
            DisplayAndInputProcessor.filterCoveredKeywords(set);
            if (set.contains("PandA beAR")) {
                errln("PandA beAR should be filtered out");
            }
        }
    }
```

It still passes. Nevertheless I'll try rewriting the method to use TreeMultimap...

btangmu commented:
@macchiati in my second commit I changed the imports to avoid asterisk (and configured my IntelliJ accordingly); I added Locale.ROOT for toLowerCase; I enlarged the new unit test to confirm "Panda and PANDA, or other cases like it"; and I added a comment to clarify what the method is intended to do.

I didn't change the implementation to use TreeMultimap as you suggested, since on further study I couldn't quite understand what you were suggesting.

This PR doesn't complete the ticket. Although it should prevent the bug that created duplicates such as "... | japanse poskantoor | Japanse poskantoor | ...", it doesn't remove such duplicates that have already been created. I intend to make another PR to accomplish that removal, and implement it with a new separate method DisplayAndInputProcessor.filterKeywordsDifferingOnlyInCase.

macchiati commented:
The advantage of avoiding * is that then you can search for files that
import a particular file; helps with maintenance.

On Thu, Aug 24, 2023 at 7:49 AM Tom Bishop ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> tools/cldr-code/src/main/java/org/unicode/cldr/test/DisplayAndInputProcessor.java
> <https://github.com/unicode-org/cldr/pull/3214#discussion_r1304457279>:
>
> > @@ -18,14 +18,7 @@
>  import com.ibm.icu.text.UnicodeSetIterator;
>  import com.ibm.icu.util.Output;
>  import com.ibm.icu.util.ULocale;
> -import java.util.Arrays;
> -import java.util.Comparator;
> -import java.util.HashMap;
> -import java.util.HashSet;
> -import java.util.List;
> -import java.util.Map;
> -import java.util.Set;
> -import java.util.TreeSet;
> +import java.util.*;
>
> evidently spotless doesn't change it one way or the other, while IntelliJ
> (default configuration) condenses to * under some circumstances -- in this
> case, when when I added ArrayList within a method and IntelliJ asked
> whether to import the class.
>
> I don't understand why one way is better than the other, other than
> brevity favoring *, but I'll try to figure out how to reconfigure IntelliJ
> if that's a problem...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unicode-org/cldr/pull/3214#discussion_r1304457279>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJLEMHGE7D5SV4NGKSSERTXW5SYXANCNFSM6AAAAAA333DMTA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

macchiati commented:
Ah, I thought this was to do both, which was why I didn't see how the code would do that. 

I think DAIP would be the best place to de-dup for case. And we can favor the lowercase value; if people want a title or uppercase, they would just replace the lowercase with that.

How the TreeMultimap would work.

Panda | PANDA | panda | BEAR | Bear
=>

 // build the TreeMultimap, sorting with the root collator
// then put lowercase(item), item

bear => {Bear, BEAR}
panda => {panda, Panda, PANDA}

Pick the first item in each value set, thus {panda, Bear}, and then join with " | ".

btangmu commented:
I accidentally made a third commit with changes for another ticket, so I reverted that change and squashed this PR to a single commit
